doc/neps/nep-0041-improved-dtype-support.rst

```diff
@@ -0,0 +1,703 @@
+If the values define an order, the category labels (internally integers) could
+be ordered the same way to allow efficient sorting and comparison.
+
+Whether or not casting is defined from one categorical with less to one with
+strictly more values defined, is something that the Categorical datatype would
+need to decide. Both options should be valid.
+
+
+Python Enums DType
+""""""""""""""""""
+
+An example for a more complex datatype, encompassing an additional concept
+would be a DType that could wrap ``enum.Enum``::
+
+    >>> class Breakfast(enum.Enum):
+    ...     spam, eggs, toast = 1, 2, '3'
+    >>> table = array([Breakfast.spam, Breakfast.eggs, Breakfast.toast], dtype=EnumDType(Breakfast))
+    >>> to_values(table)  # to_values is a ufunc here
+    array([1, 2, '3'], dtype=object)
+    >>> table[0]
+    <Breakfast.spam: 1>
+
+The following operations may be desirable but are *not* straight forward::
+
+    >>> table2 = np.array([Breakfast.spam, Breakfast.eggs])  # discover same dtype
+    >>> table == Breakfast.spam
+    array([True, False, False])
+
+To define these NumPy would need to find the correct Enum DType.
+And it is unclear how that should be done, because the type associated with
+an ``EnumDType`` could be any ``Enum`` subclass.
+
+An alternative approach, which solves those issues, is to create the DTypes as::
+
+    >>> BreakfastDType = EnumDType(Breakfast)  # class/DType factory
+    >>> issubclass(BreakfastDType, EnumDType)
+    True
+
+Where ``BreakfastDType`` is a subclass and not just an instance of ``EnumDType``.
+Which makes sense since also ``Enum`` is a class factory.
+The Enum example is particularly complicated (it may be good to keep in mind,
+but it is likely that is would not be supported initially).
+
+
+Unit on the Datatype
+""""""""""""""""""""
+
+There are different ways to define Units, depending on how the internal
+machinery would be organized, one way is to have a single Unit datatype
+for every existing numerical type.
+This will be written as ``Unit[float64]``, the unit itself is part of the
+DType instance ``Unit[float64]("m")`` us a ``float64`` with meters attached::
+
+    >>> meters = np.array([1, 2, 3], dtype=np.float64) * unit.m  # meters
+    >>> print(meters)
+    array([1.0, 2.0, 3.0], dtype=Unit[float64]("m"))
+
+Note that units are a bit tricky, since it is debatable, whether::
+
+    >>> np.array([1.0, 2.0, 3.0], dtype=Unit[float64]("m"))
+
+should be valid syntax (coercing the float scalars without a unit to meters).
+Once the array is created, math will work without any issue::
+
+    >>> meters * 2 * unit.seconds
+    array([2.0, 4.0, 6.0], dtype=Unit[float64]("ms"))
+
+Casting is not valid from one unit to the other, but can be between different
+scales of the same dimensionality (although this may usually be "unsafe")::
+
+    >>> meters.astype(Unit[float64]("s"))
+    TypeError: Cannot cast meters to seconds.
+    >>> meters.astype(Unit[float64]("km"))
```

Comment:
On second read, it's not very clear to me how this would work under the hood.
Does it imply that the Unit class will need to define something like a magic method
```python
def __astype__(self, array: np.ndarray, newtype: numpy.DType) -> np.ndarray:
```
?
What happens if the two types belong to different libraries? Would it be possible to define
```python
def __fromtype__(self, array: np.ndarray) -> np.ndarray:
```
in the target dtype if the ``__astype__`` before in library A (which defined array.dtype) raises TypeError?

Reply:
They would provide `CastingImpl` objects (or maybe rather C-structs with a specific, extensible initialization syntax) which houses all the necessary information, mainly a way to check casting safety and "adjust" the dtypes and then an actual loop doing the cast. These things are in NEP 42, although that will needs a bit settling on the API and revising on the text, I think the general notions are sound by now.

That can be extended later (similar to python which has `__getattr__` with which you can get an object even though it is not part of the class dictionary), but initially I want you to provide a list of possible casts at creation time: I.e. one of the DTypes must know about the other one.

I think that is a good "basic" design choice, because I do not really want users to discourage messing with the casting of datatypes you did not define.